[DOCS-1410] Updates to smoothing algorithm docs (#1270)
wandb/docs: content/guides/models/app/features/panels/line-plot/smoothing.md

weight: 30
---

W&B supports several types of smoothing:

- [Time weighted exponential moving average]({{< relref "#time-weighted-exponential-moving-average-default" >}}) 

> - [gaussian smoothing]({{< relref "#gaussian-smoothing" >}}) (default)
> - [running average]({{< relref "#running-average" >}})

Author: ```suggestion
- [Gaussian smoothing]({{< relref "#gaussian-smoothing" >}}) (default)
- [Running average]({{< relref "#running-average" >}})
```

weight: 30
---

> W&B supports three types of back-end smoothing applied by W&B Server:

Reviewer: ```suggestion
W&B supports three types of back-end smoothing applied by W&B Server:
```

Only the exponential moving average smoothing algorithm (marked tensorboard below) has backend support as of my change. The other three types all are done exclusively on the frontend. The new feature I added is the backend support for EMA

Author: Done, PTAL

See these live in an [interactive W&B report](https://wandb.ai/carey/smoothing-example/reports/W-B-Smoothing-Features--Vmlldzo1MzY3OTc).

{{< img src="/images/app_ui/beamer_smoothing.gif" alt="" >}}

## Time Weighted Exponential Moving Average (Default)

The Time Weighted Exponential Moving Average (TWEMA) smoothing algorithm is a technique for smoothing time series data by exponentially decaying the weight of previous points. The range is 0 to 1. See [Exponential Smoothing](https://www.wikiwand.com/en/Exponential_smoothing) for background. There is a de-bias term added so that early values in the time series are not biased towards zero.

The TWEMA algorithm takes the density of points on the line (the number of `y` values per unit of range on x-axis) into account. This allows consistent smoothing when displaying multiple lines with different characteristics simultaneously.

TWEMA smoothing is applied by W&B Server to a full scan of history, rather than sampling or bucketing first before smoothing. In general, it is more accurate than EMA smoothing. However, in the following situations, TWEMA smoothing is not supported and EMA smoothing is used instead:
- Sampling
- Grouping

> - non-monotonic x-axes

Reviewer: This only applies to the exponential moving average smoothing algorithm not TWEMA

Author: Done, PTAL

Here's what this looks like [in the app](https://wandb.ai/carey/smoothing-example/reports/W-B-Smoothing-Features--Vmlldzo1MzY3OTc):

{{< img src="/images/app_ui/weighted_exponential_moving_average.png" alt="Demo of TWEMA smoothing" >}}

## Gaussian smoothing

Gaussian smoothing (or gaussian kernel smoothing) computes a weighted average of the points, where the weights correspond to a gaussian distribution with the standard deviation specified as the smoothing parameter. The smoothed value is calculated for every input x value.

> Gaussian smoothing is a good standard choice for smoothing if your points or spaced unevenly on the x-axis, or if you are not concerned with matching a different smoothing behavior. Unlike with TWEMA or EMA, the point will be smoothed based on points occurring both before and after the value.

Reviewer 2: Do we need this guidance?  Also people can and do use Gaussian smoothing to match behavior in other tools' Gaussian smoothing implementations so I don't think that's relevant to say here.

Author: I removed most of this paragraph and moved the "before and after" detail up into the previous paragraph.

## Running average smoothing

Running average is a smoothing algorithm that replaces a point with the average of points in a window before and after the given x value. See "Boxcar Filter" at [https://en.wikipedia.org/wiki/Moving_average](https://en.wikipedia.org/wiki/Moving_average). The selected parameter for running average tells Weights and Biases the number of points to consider in the moving average.

Consider using Gaussian Smoothing instead if your points are spaced unevenly on the x-axis.

Here's what this looks like [in the app](https://wandb.ai/carey/smoothing-example/reports/W-B-Smoothing-Features--Vmlldzo1MzY3OTc#4.-running-average):

{{< img src="/images/app_ui/running_average.png" alt="Demo of running average smoothing" >}}

## Exponential Moving Average (EMA) smoothing

The Exponential Moving Average (EMA) smoothing algorithm is a rule of thumb technique for smoothing time series data using the exponential window function. For details about the technique, see [Exponential Smoothing](https://www.wikiwand.com/en/Exponential_smoothing). The range is 0 to 1. A debias term is added so that early values in the time series are not biases towards zero.

> On a best-effort basis, EMA smoothing is applied at the back end by W&B Server. In this situation, EMA smoothing is applied to a full scan of history, rather than bucketing first before smoothing. In general, this improves the accuracy of the smoothing operation.

Reviewer 2: I would delete this sentence.  Best-effort presents the product is deficient, back end should be one word or hyphenated (?), and this is not what W&B Server means I think.

Author: I think that Danny thought it was important that the smoothing would happen to the full scan if it's handled at the back end, and to bucketed data otherwise.  PTAL at the latest attempt.

In the following situations, EMA smoothing is instead applied at the front end by the W&B App, after bucketing.
- Sampling
- Grouping
- Expressions
- Non-monotonic x-axes
- Time-based x-axes

> Here is sample code for how this works under the hood:

Reviewer 2: Why do we need this?

Author: I didn't add this. Would you like me to remove the code block? How about the code block for TWEMA smoothing above?

The following image demonstrates how a running app looks like [in the app](https://wandb.ai/carey/smoothing-example/reports/W-B-Smoothing-Features--Vmlldzo1MzY3OTc#4.-running-average):

{{< img src="/images/app_ui/running_average.png" alt="Demo of running average smoothing" >}}

## Exponential Moving Average

> The Exponential Moving Average (EMA) smoothing algorithm is implemented to match TensorBoard's smoothing algorithm. The range is 0 to 1. See [Exponential Smoothing](https://www.wikiwand.com/en/Exponential_smoothing) for background. A debias term is added so that early values in the time series are not biases towards zero.

Reviewer 2: I would remove any reference to Tensorboard, I don't think that's true of the new version, or required.

Author: Done, PTAL. This required some additional adjustments, and then I noticed a few other areas of inconsistency and fixed those too, in the latest commit.

On a best-effort basis, EMA smoothing is applied at the back end by W&B Server. In this situation, EMA smoothing is applied to a full scan of history, rather than bucketing first before smoothing. In general, this improves the accuracy of the smoothing operation.

In the following situations, EMA smoothing is instead applied at the front end by the W&B App, after bucketing.
- Sampling
- Grouping
- Expressions
- Non-monotonic x-axes

> - time-based x-axes

Author: ```suggestion
- Time-based x-axes
```

The following image demonstrates how a running app looks like [in the app](https://wandb.ai/carey/smoothing-example/reports/W-B-Smoothing-Features--Vmlldzo1MzY3OTc#4.-running-average):

{{< img src="/images/app_ui/running_average.png" alt="Demo of running average smoothing" >}}

## Exponential Moving Average

The Exponential Moving Average (EMA) smoothing algorithm is implemented to match TensorBoard's smoothing algorithm. The range is 0 to 1. See [Exponential Smoothing](https://www.wikiwand.com/en/Exponential_smoothing) for background. A debias term is added so that early values in the time series are not biases towards zero.

EMA smoothing is applied by the W&B App after sampling and bucketing. In general, it is less accurate than [TWEMA smoothing]({{< relref "#time-weighted-exponential-moving-average-default" >}}). However, EMA is used instead of TWEMA in the following situations:
- Sampling
- Grouping

> - non-monotonic x-axes

Reviewer: this list should also include 

- expressions
- time-based x axes

Author: Done

Here's what this looks like [in the app](https://wandb.ai/carey/smoothing-example/reports/W-B-Smoothing-Features--Vmlldzo1MzY3OTc):

{{< img src="/images/app_ui/exponential_moving_average.png" alt="Demo of EMA smoothing" >}}

## Implementation Details

> With the exception of back-end EMA smoothing, all of the smoothing algorithms run on the sampled data. If you log more than 1500 points, the smoothing algorithm will run _after_ the points are downloaded from W&B Server. The intention of the smoothing algorithms is to help find patterns in data quickly. If you need exact smoothed values on metrics with a large number of logged points, it may be better to download your metrics through the API and run your own smoothing methods.

Author: Is this paragraph still accurate?

> With the exception of back-end EMA smoothing, all of the smoothing algorithms run on the sampled data. If you log more than 1500 points, the smoothing algorithm will run _after_ the points are downloaded from W&B Server. The intention of the smoothing algorithms is to help find patterns in data quickly. If you need exact smoothed values on metrics with a large number of logged points, it may be better to download your metrics through the API and run your own smoothing methods.

Reviewer: ```suggestion
With the exception of back-end EMA smoothing, all of the smoothing algorithms run on the sampled or binned data (if full fidelity mode is enabled). If you log more than 1500 points, the smoothing algorithm will run _after_ the points are downloaded from W&B Server. The intention of the smoothing algorithms is to help find patterns in data quickly. If you need exact smoothed values on metrics with a large number of logged points, it may be better to download your metrics through the API and run your own smoothing methods.
```

> With the exception of back end EMA smoothing, all of the smoothing algorithms run on the sampled data. If you log more than 1500 points, the smoothing algorithm will run _after_ the points are downloaded from W&B Server. The intention of the smoothing algorithms is to help find patterns in data quickly. If you need exact smoothed values on metrics with a large number of logged points, it may be better to download your metrics through the API and run your own smoothing methods.

Reviewer 2: `the points are downloaded from W&B Server`  This is not how we refer to W&B Server which is a product.  "Returned from the database" might be more accurate.  Broadly I don't think this section on implementation details is required at all

Author: Removed. It was there before my PR.

## Hide original data

> By default, the original, unsmoothed data displays in the plot as a faint line in the background. Click the **Show Original** toggle to turn this off.

Author: ```suggestion
By default, the original unsmoothed data displays in the plot as a faint line in the background. Click **Show Original** to turn this off.
```

Here's what this looks like [in the app](https://wandb.ai/carey/smoothing-example/reports/W-B-Smoothing-Features--Vmlldzo1MzY3OTc):

{{< img src="/images/app_ui/exponential_moving_average.png" alt="Demo of EMA smoothing" >}}

## Implementation Details

> With the exception of TWEMA, all of the smoothing algorithms run on the sampled data. If you log more than 1500 points, the smoothing algorithm will run _after_ the points are downloaded from W&B Server. The intention of the smoothing algorithms is to help find patterns in data quickly. If you need exact smoothed values on metrics with a large number of logged points, it may be better to download your metrics through the API and run your own smoothing methods.

Reviewer: ```suggestion
With the exception of EMA, all of the smoothing algorithms run on the sampled or binned data. If you log more than 1500 points, the smoothing algorithm will run _after_ the points are downloaded from W&B Server. The intention of the smoothing algorithms is to help find patterns in data quickly. If you need exact smoothed values on metrics with a large number of logged points, it may be better to download your metrics through the API and run your own smoothing methods.
```
